ambry-store/src/main/java/com/github/ambry/store/HybridCompactionPolicy.java

```diff
@@ -0,0 +1,191 @@
+import com.github.ambry.utils.Time;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.codehaus.jackson.map.ObjectMapper;
+
+
+/**
+ * HybridCompactionPolicy will run compaction with StatsBasedCompactionPolicy more frequently and with CompactAllPolicy
+ * one time out of storeCompactionPolicySwitchPeriod.
+ */
+public class HybridCompactionPolicy implements CompactionPolicy {
+  private final Time time;
+  private final StoreConfig storeConfig;
+  private static final Logger logger = LoggerFactory.getLogger(HybridCompactionPolicy.class);
+  private final Map<String, CompactionPolicySwitchInfo> blobToCompactionPolicySwitchInfoMap;
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+  private static final String COMPACT_POLICY_INFO_PATH = File.separator + "compactionPolicyInfo.json";
+  private static final int INIT_COMPACT_ALL_TIME = 0;
+  private static final int INIT_COUNTER_VALUE = 0;
+
+  HybridCompactionPolicy(StoreConfig storeConfig, Time time) {
+    this.storeConfig = storeConfig;
+    this.time = time;
+    this.blobToCompactionPolicySwitchInfoMap = new HashMap<>();
+  }
+
+  /**
+   * @param totalCapacity Total capacity of the {@link BlobStore}
+   * @param usedCapacity Used capacity of the {@link BlobStore}
+   * @param segmentCapacity Segment capacity of a {@link LogSegment}
+   * @param segmentHeaderSize Segment header size of a {@link LogSegment}
+   * @param logSegmentsNotInJournal {@link List<String> } of log segment names which has non overlapping entries with
+   *                                {@link Journal}
+   * @param blobStoreStats {@link BlobStoreStats} pertaining to the {@link BlobStore} for which
+   * {@link CompactionDetails} are requested
+   * @param dataDir the dir to store {@link CompactionPolicySwitchInfo}
+   * @return {@link CompactAllPolicy} or {@link StatsBasedCompactionPolicy}'s {@link CompactionDetails} depends on the switching rules.
+   */
+  @Override
+  public CompactionDetails getCompactionDetails(long totalCapacity, long usedCapacity, long segmentCapacity,
+      long segmentHeaderSize, List<String> logSegmentsNotInJournal, BlobStoreStats blobStoreStats, String dataDir)
+      throws StoreException {
+    String storeId = blobStoreStats.getStoreId();
+    CompactionPolicySwitchInfo compactionPolicySwitchInfo = getCompactionPolicySwitchInfo(storeId, dataDir, blobStoreStats);
+    CompactionPolicy selectCompactionPolicy =
+        selectCompactionPolicyAndUpdateCompactionPolicySwitchInfo(compactionPolicySwitchInfo);
+    logger.info("Current compaction policy  is : {}", selectCompactionPolicy);
+    backUpCompactionPolicyInfo(dataDir, compactionPolicySwitchInfo);
+    compactionPolicySwitchInfo.getCompactionPolicyCounter().increment();
```

Comment:
shouldn't we increment the counter before select which policy to use?

Reply:
That's the tricky part. I'm planning to increment after the backup with two reasons:
1. when counter equals to 0, it will run compactAllPolicy. which means if we don't have any back up files yet, we will run compact all policy.
2. if we are running compactAll for a long time and host restart, the file backs up the current status and re-run compactAll instead of statsBased.
Let me know if you have any other concern.